stdlib/private/OSLog/OSLogMessage.swift

```diff
@@ -0,0 +1,285 @@
+  /// All arguments to this function must be known at compile time.
+  public func getIntegerFormatSpecifier(
+    _ format: IntFormat,
+    isPrivate: Bool,
+    isLong: Bool,
+    isUnsigned: Bool)
+    -> String {
+    var formatSpecifier: String = isPrivate ? "%{private}" : "%{public}"
+    formatSpecifier += isLong ? "l" : ""
+
+    switch (format) {
+    case .hex:
+      formatSpecifier += "x"
+    case .octal:
+      formatSpecifier += "o"
+    default:
+      formatSpecifier += isUnsigned ? "u" : "d"
+    }
+    return formatSpecifier;
+  }
+}
+
+extension String {
+  /// Replace all percents "%" in the string by "%%" so that the string can be
+  /// interpreted as a C format string.
+  public var percentEscapedString: String {
+    get {
+      return self.reduce(into: "") { (result, char) in
+        result += (char == "%") ? "%%" : String(char)
+      }
+    }
+  }
+}
+
+public struct OSLogMessage :
+  ExpressibleByStringInterpolation, ExpressibleByStringLiteral
+{
+  let oslogInterpolation : OSLogInterpolation
+
+  /// Initializer for accepting string interpolations.
+  public init(stringInterpolation: OSLogInterpolation) {
+    oslogInterpolation = stringInterpolation
+  }
+
+  /// Initializer for accepting string literals.
+  public init(stringLiteral value: String) {
+    // Note that the actual value of `literalCapacity` is not important as it
+    // is ignored by `OSLogInterpolation.init`. However, it must be a literal.
+    var s = OSLogInterpolation(literalCapacity: 1, interpolationCount: 0)
+    s.appendLiteral(value)
+    self.init(stringInterpolation: s)
+  }
+
+  /// Format string constructed from the string interpolation.
+  public var formatString: String {
+    get { return oslogInterpolation.formatString }
+  }
+
+  /// The byte size of the argument buffer that will contain the preamble
+  /// and the elements of `oslogInterpolation.argumentSequence`.
+  public var argumentBufferByteSize: Int {
+    get { return oslogInterpolation.argumentSequence.byteCount + 2 }
+  }
+
+  /// Serialize the summary bytes and arguments into the given byte-buffer
+  /// builder.
+  public func serializeArguments(
+    into bufferBuilder: inout OSLogByteBufferBuilder) {
+    bufferBuilder.serialize(oslogInterpolation.preamble)
+    bufferBuilder.serialize(oslogInterpolation.argumentCount)
+    oslogInterpolation.argumentSequence.serialize(into: &bufferBuilder)
+  }
+}
+
+/// A representation of a sequence of arguments and headers (of possibly
+/// different types) that have to be serialized to a byte buffer. The arguments
+/// are captured within closures and stored in an array. The closures accept an
+/// instance of `OSLogByteBufferBuilder`, and when invoked, serialize the
+/// argument using the passed `OSLogByteBufferBuilder` instance.
+public struct OSLogArgumentSequence {
+  /// An array of closures that captures arguments of possibly different types.
+  public var argumentClosures: [(inout OSLogByteBufferBuilder) -> ()]
```

Comment:
Does it need to be public?

Reply:
The idea to allow this struct to be extended with overloads of `append` for custom types. Those extensions would have to add elements to this array, that unfortunately needs this property to be mutable from outside.